components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/PingScheduler.kt

```diff
@@ -0,0 +1,213 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean.scheduler
+
+import android.arch.lifecycle.Lifecycle
+import android.arch.lifecycle.LifecycleObserver
+import android.arch.lifecycle.OnLifecycleEvent
+import android.arch.lifecycle.ProcessLifecycleOwner
+import android.content.Context
+import android.support.annotation.VisibleForTesting
+import androidx.work.BackoffPolicy
+import androidx.work.Constraints
+import androidx.work.NetworkType
+import androidx.work.OneTimeWorkRequestBuilder
+import androidx.work.OneTimeWorkRequest
+import androidx.work.WorkInfo
+import androidx.work.WorkManager
+import mozilla.components.service.glean.Glean
+import mozilla.components.service.glean.GleanMetrics.GleanBaseline
+import mozilla.components.service.glean.ping.BaselinePing
+import mozilla.components.service.glean.ping.PingMaker
+import mozilla.components.service.glean.storages.PingStorageEngine
+import mozilla.components.service.glean.storages.StorageEngineManager
+import mozilla.components.support.base.log.logger.Logger
+import java.util.UUID
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.TimeUnit
+
+/**
+ * This class is responsible for scheduling of glean pings. In order to do this in association
+ * with lifecycle events, it implements [LifecycleObserver].  On entering the background
+ * [Lifecycle.Event.ON_STOP] state, it will use [WorkManager] to schedule a [PingUploadWorker] task
+ * which is responsible for uploading the ping.
+ */
+internal class PingScheduler(
+    applicationContext: Context,
+    storageEngineManager: StorageEngineManager,
+    pingMaker: PingMaker = PingMaker(storageEngineManager, applicationContext),
+    pingStorageEngine: PingStorageEngine = PingStorageEngine(applicationContext),
+    metricsPingScheduler: MetricsPingScheduler = MetricsPingScheduler(applicationContext)
+) : LifecycleObserver {
+    private val logger = Logger("glean/GleanLifecycleObserver")
+
+    // The application id detected by glean to be used as part of the submission
+    // endpoint.
+    internal var applicationId = sanitizeApplicationId(applicationContext.packageName)
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val pingMaker: PingMaker = pingMaker
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val pingStorageEngine = pingStorageEngine
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    val metricsPingScheduler = metricsPingScheduler
```

Comment:
nit: You can mark them as `val` inside the constructor above. (maybe `internal` too?)